### DISCUSSION TITLE
[Request] Channels DVR Server

### DISCUSSION THREAD

**Discussion by kisonay** [tteck/Proxmox#2126] [Ideas]
Would love to see a setup for Channels DVR Server.

A flexible, standalone media server that lets you watch tv your way.

https://getchannels.com/dvr-server/

https://getchannels.com/dvr-server/#linux

**Comment by tteck** [comment]
Interesting

**Reply by kisonay** [reply]
I'm just starting with Proxmox (migrating from a pure docker setup) but it seems pretty straight forward (similar to Plex or jellyfin.) Would love to help in anyway I can.

**Reply by di5cord20** [reply]
I haven't used Channels DVR before.  Looks interesting.  Would you say it's a Plex or Jellyfin replacement?

**Reply by kisonay** [reply]
Neither. I have been using it alongside Plex for years now. It’s specifically geared towards TV (with TV Anywhere support ie recording live streams from the different networks apps)

**Reply by di5cord20** [reply]
Are you using the subscription or HDHomeRun only solution (which I believe doesn't have recording capabilities)?

**Reply by kisonay** [reply]
HDHomeRun is not required but I do use it with both HDHomeRun and TV Everywhere (through Philo)

I do subscribe as I enjoy the product and it replaced my TiVo.

I understand if this doesn't fall into the proper category for this repo. I have been playing with a Debian LXC and installing it there.

**Comment by tteck** [comment]
Ready for testing

```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/channels.sh)"
```
Please let me know if we want this permanently added to the repository.

**Reply by tteck** [reply]
Currently, "Hardware Acceleration Support" has **NOT** been added.

**Reply by kisonay** [reply]
I’ll be able to test tomorrow.

**Reply by di5cord20** [reply]
I can test (although I've never used Channels before).  My assumption is the script should run on the VE shell, not LXC... is that correct?

**Reply by tteck** [reply]
Correct

**Reply by di5cord20** [reply]
Just ran using Advanced but accepting all defaults.  

![image](https://github.com/tteck/Proxmox/assets/140566995/de609e2a-c368-4961-aefb-09ff5f8babac)

**Reply by kisonay** [reply]
Took a quick look and installed it on a test Proxmox VM.

<img width="918" alt="Screenshot 2023-11-30 at 6 22 00 PM" src="https://github.com/tteck/Proxmox/assets/1854267/c747a92f-4aa6-4e91-8602-f159ba4c53df">

Initial test looks good. Starts up without a problem and I can configure it with my HDHR. Would recommend changing default RAM to 1024 (per their documentation.)

Tomorrow I will install it on my "live" instance and import a backup from when I was running it in a Docker Container. When I do that, I can also install the required packages for HW transcoding and modify the config to mount dev/dri and see if I can get it working.

**Reply by kisonay** [reply]
actually one thing that will be a blocker (now that I took a look at the install scripts.)

can we please install the TVE dependencies by default (or have an option during install to include them)?

apt-get install chromium-browser xvfb

**Reply by tteck** [reply]
Added extra dependencies and adjusted RAM

To use hardware acceleration, execute this in Proxmox shell and select the container

```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/misc/hw-acceleration.sh)"
```

**Reply by tteck** [reply]
use the default Container Type "0"  `CT_TYPE="0"` (privileged) for hardware acceleration.

**Reply by kisonay** [reply]
Doing my testing and it appears the added dependencies are breaking on Debian12. 
Ubuntu uses "chromium-browser" while Debian uses "chromium"

```
Package chromium-browser is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'chromium-browser' has no installation candidate
++ error_handler 20 '$STD apt-get install -y chromium-browser'
++ local exit_code=100
++ local line_number=20
++ local 'command=$STD apt-get install -y chromium-browser'
++ local 'error_message=\033[01;31m[ERROR]\033[m in line \033[01;31m20\033[m: exit code \033[01;31m100\033[m: while executing command \033[33m$STD apt-get install -y chromium-browser\033[m'
++ echo -e '\n\033[01;31m[ERROR]\033[m in line \033[01;31m20\033[m: exit code \033[01;31m100\033[m: while executing command \033[33m$STD apt-get install -y chromium-browser\033[m'
```
Since Debian12 is default, can we update to "chromium"?

**Reply by kisonay** [reply]
Working great! Imported my backup without issue and up and running. Thank you! this allows me to easily replace my docker setup.

**Reply by stuartisaac** [reply]
Could you add the info about enabling hardware acceleration (using a privileged container and running the script) to the main section for Channels (e.g. below the `Channels DVR Server Interface: IP:8089` line)? It took me a bit to realize why hardware acceleration wasn't working until I found this discussion. 
Thanks!

**Reply by tteck** [reply]
Hardware Acceleration is now added by default.

**Comment by tteck** [comment]
Added to the website.

**Comment by Vakrehus** [comment]
This only works with a subscription to save **MY** local stream from **MY** device in **MY** living room?
Puhh...so I pay twice for my cable connection.

**Reply by kisonay** [reply]
This is not for everyone and isn't an "addon" in my opinion. This is not for people with a cable subscription, it is for people looking to cut cable and have a DVR with antenna. Has the added benefit of TV Everywhere if you subscribe to one of the cheaper services that support it.

**Comment by DrD00der** [comment]
Running the script I get the following error:
```
[ERROR] in line 23: exit code 0: while executing command "$@" > /dev/null 2>&1
The silent function has suppressed the error, run the script with verbose mode enabled, which will provide more detailed output.
```

It does create the container but I can't seem to access it.  I can't access the server at the IP address of the container.  If I open console, it asks for a channels login/password.  I can't seem to figure out what they are to log in properly.  I tried my promox root/password and my channels DVR login/password.  
Any suggestions? 
Thanks!

**Reply by DrD00der** [reply]
Update: I ran the script in verbose mode and got a more detailed error log:
`E: Failed to fetch http://deb.debian.org/debian/pool/main/g/gtk%2b3.0/libgtk-3-common_3.24.24-4%2bdeb11u3_all.deb  Hash Sum mismatch
   Hashes of expected file:
    - SHA256:d37b9524bcb56149fe2b5f61618e962c6098f0d71c64c97c5047a652cc1c935e
    - MD5Sum:efd0efc3068db3dc8d06e0c9800db047 [weak]
    - Filesize:3757484 [weak]
   Hashes of received file:
    - SHA256:49ccc3c0e87997fad361e7a835ebd52d9ba9f53ec268bb96820b102fedfe894a
    - MD5Sum:1daab8a4d208a208362476151043fcf3 [weak]
    - Filesize:3757484 [weak]
   Last modification reported: Fri, 07 Apr 2023 14:55:40 +0000
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
|
[ERROR] in line 20: exit code 0: while executing command $STD apt-get install -y chromium`

**Reply by tteck** [reply]
What Linux distribution are you trying to use?

**Reply by DrD00der** [reply]
debian

**Reply by tteck** [reply]
Let me test...

**Reply by tteck** [reply]
The script runs without any problems from this end.
```
   ________                           __        ____ _    ______     _____
  / ____/ /_  ____ _____  ____  ___  / /____   / __ \ |  / / __ \   / ___/___  ______   _____  _____
 / /   / __ \/ __ `/ __ \/ __ \/ _ \/ / ___/  / / / / | / / /_/ /   \__ \/ _ \/ ___/ | / / _ \/ ___/
/ /___/ / / / /_/ / / / / / / /  __/ (__  )  / /_/ /| |/ / _, _/   ___/ /  __/ /   | |/ /  __/ /
\____/_/ /_/\__,_/_/ /_/_/ /_/\___/_/____/  /_____/ |___/_/ |_|   /____/\___/_/    |___/\___/_/

Using Advanced Settings
Using Distribution: debian
Using debian Version: 12
Using Container Type: 0
Using Root Password: Automatic Login
Container ID: 110
Using Hostname: channels
Using Disk Size: 8
Allocated Cores: 2
Allocated RAM: 1024
Using Bridge: vmbr0
Using IP Address: dhcp
Using Gateway IP Address: Default
Using APT-Cacher IP Address: Default
Disable IPv6: yes
Using Interface MTU Size: Default
Using DNS Search Domain: Host
Using DNS Server IP Address: Host
Using MAC Address: AE:1A:60:07:B7:59
Using Vlan: Default
Enable Root SSH Access: no
Enable Verbose Mode: no
Creating a Channels LXC using the above advanced settings
 ✓ Using local for Template Storage.
 ✓ Using thin-pool for Container Storage.
 ✓ Updated LXC Template List
 ✓ LXC Container 110 was successfully created.
 ✓ Started LXC Container
 ✓ Set up Container OS
 ✓ Network Connected: 192.168.86.250 
 ✓ IPv4 Internet Connected
 ✗ IPv6 Internet Not Connected
 ✓ DNS Resolved github.com to 140.82.112.4
 ✓ Updated Container OS
 ✓ Installed Dependencies
 ✓ Set Up Hardware Acceleration
 ✓ Installed Channels DVR Server
 ✓ Customized Container
 ✓ Cleaned
 ✓ Completed Successfully!

Channels Setup should be reachable by going to the following URL.
         http://192.168.86.250:8089  
```

**Reply by DrD00der** [reply]
Hmm, I'm struggling to figure out why this isn't working for me.  I also get lots of random errors on other things that should work.  I get an error for the Scrypted script install as well.  The ubuntu ones work fine on my end.  Even did a fresh install of Proxmox and wiped the whole box (Beelink SEi12).  Any suggestion on what else I could look at?

**Reply by tteck** [reply]
Show the script output.

**Reply by DrD00der** [reply]
welp, mysteries of the world...
I've tried this multiple times and kept getting errors.  Just now was able to run both Channels and Scrypted scripts in verbose mode without errors.  Both threw errors when not in verbose mode, but now both are installed and working.  

Wish I knew why that would be the case, but glad it's working now. 

Thanks!

**Reply by tteck** [reply]
sounds like cache issues

**Comment by stuartisaac** [comment]
I decided to try setting up channels in an unprivileged container, and found that it worked, including hardware acceleration (passing through `/dev/dri/card1` and `/dev/dri/renderD128`).  I did so by creating a plain Debian container, adding a user to do UID mapping with the host, adding the user to `sudo` and the `video` and `render` groups as indicated on the channels website, then running the installer script from the channels website. One annoyance was that Channels seems to look for `/dev/dri/card0`, which at some point changed to `/dev/dri/card1` on my system for some reason, and I'm working around that by symlinking card0 to card1 using a script called from `lxc.hook.start`. This seems to work, but maybe there's a better way of resolving it?

Also, @tteck, I was wondering if you would consider updating the channels script to use an unprivileged container in this way? Or maybe the step of creating a non-root user is a little much?

**Reply by tteck** [reply]
If the Channels LXC is created unprivileged, you still get "Hardware Acceleration Support"

![Screenshot 2024-06-15 7 33 48 AM](https://github.com/tteck/Proxmox/assets/21342670/9dbcb372-9d76-4afe-b983-9a04da766539)

**Reply by stuartisaac** [reply]
I see. Is there some benefit to having the container be privileged by default, then?
